marlowe-playground-client/src/Marlowe/Extended.purs

```diff
@@ -534,3 +703,23 @@ instance toCoreContract :: ToCore Contract S.Contract where
   toCore (When cases tim cont) = S.When <$> traverse toCore cases <*> toCore tim <*> toCore cont
   toCore (Let varId val cont) = S.Let <$> pure varId <*> toCore val <*> toCore cont
   toCore (Assert obs cont) = S.Assert <$> toCore obs <*> toCore cont
+
+instance templateContract :: Template Contract Placeholders where
+  getPlaceholderIds Close = mempty
+  getPlaceholderIds (Pay accId payee tok val cont) = getPlaceholderIds val <> getPlaceholderIds cont
+  getPlaceholderIds (If obs cont1 cont2) = getPlaceholderIds obs <> getPlaceholderIds cont1 <> getPlaceholderIds cont2
+  getPlaceholderIds (When cases tim cont) = foldMap getPlaceholderIds cases <> getPlaceholderIds tim <> getPlaceholderIds cont
+  getPlaceholderIds (Let varId val cont) = getPlaceholderIds val <> getPlaceholderIds cont
+  getPlaceholderIds (Assert obs cont) = getPlaceholderIds obs <> getPlaceholderIds cont
+
+instance fillableContract :: Fillable Contract TemplateContent where
+  fillTemplate placeholders contract = case contract of
+    Close -> Close
+    Pay accId payee tok val cont -> Pay accId payee tok (go val) (go cont)
+    If obs cont1 cont2 -> If (go obs) (go cont1) (go cont2)
+    When cases tim cont -> When (map go cases) (go tim) (go cont)
+    Let varId val cont -> Let varId (go val) (go cont)
+    Assert obs cont -> Assert (go obs) (go cont)
+    where
+    go :: forall a. (Fillable a TemplateContent) => a -> a
+    go = fillTemplate placeholders
```

Comment:
maybe it's a waste of time but when we have three functions with the same recursion pattern is it time for a recursion scheme? All three here would be catamorphisms.

Reply:
I think it's not worth it but it is worth thinking about at least

Reply:
I considered it, but I am not sure how to do it. I did look at it last time you suggested but still didn't get it

Reply:
http://shmish111.github.io/2019/04/13/recursion-schemes-patterns/ 😉 

The only reason I didn't have this in core marlowe is to keep the semantics unchaned from the haskell version but since we don't use the semantics in extended marlowe it might be worth it. Basically it is abstracting all the calls to `go` in this function and the same for the other types.

Reply:
Nice entry ;). I've read about recursion schemes before but never actually implemented. I'm keen to give it a try next time we refactor this code. But to check if it's worth it, how many recursive implementations we have that can use this? Also, could the BlockToTerm that I wrote in #2728 be refactored with this? And finally, how would you get rid of the concrete type and just have the `ContractF`?

Reply:
any time you have a recursive data type you can pull out the recursion and then write algorithms that deal with that recursion in a specic way, the simplest (to me) being a catamorphism (see my article above). I don't know if there is a recursion scheme for a monadic catamorphism but I would guess that there is and I think that's what you would need for `BlockToTerm`.
Whether it's worth it or not is subjective. If we don't ever need the concrete types (which if you go for it you probably won't) then you don't really add any boiler plate. However while you are migrating or if you need the old data type exactly as it is then you will end up with quite a bit of boiler plate. So it's a bit like lenses, does the team want to go for it. I say yes 😉

Reply:
I'm going to put a pin on this for now :)